### PR TITLE
Ctest

### PR DESCRIPTION
**Pull request by paskino**
Uses CTest to run the tests that are found in the SIRF/src/x*/p*/tests directories.
Creates the .sirfrc file which can be sourced to set up all the environmental variables needed to run SIRF.

### CODE REVIEW

**Comment by paskino** [comment]
The build doesn't work because of these errors that are unrelated to the commits:

> The command "pip${PYVER} install --upgrade pip setuptools wheel" failed and exited with 127 during .
> The command "pip${PYVER} install --upgrade pip setuptools wheel" failed and exited with 2 during .

**Comment by KrisThielemans** [comment]
the python2 version fails because of user permissions. seems we need
```
pip${PYVER} install --user --upgrade pip setuptools wheel
```
The python3 version fails with
```
$ pip${PYVER} install --upgrade pip setuptools wheel
pyenv: pip3: command not found
The `pip3' command exists in these Python versions:
  3.5
  3.5.3
```
Possibly we need
```
 PYENV_VERSION=3.5 
```
or something, although that's a bit horrible as it depends on Travis. 
see e.g.
https://github.com/praekeltfoundation/travis-pyenv/blob/develop/.travis.yml

**Comment by casperdcl** [comment]
Hopefully #33 fixes this issue

**Comment by KrisThielemans** [comment]
@paskino, please merge with master to get @casperdcl's fix to .travis.yml

**Comment by paskino** [comment]
It now builds correctly.

File: SuperBuild.cmake [hunk 1]
```diff
@@ -92,3 +92,31 @@ set(SIRF_SRC_PATH ${CMAKE_CURRENT_LIST_DIR}/SIRF)
...
+else()
+  # create .sirfrc
+  configure_file(env_ccppetmr.sh.in $ENV{HOME}/.sirfrc)
+  # append .sirfrc to .bashrc
+#  file(APPEND $ENV{HOME}/.bashrc "#Environment variables for SIRF
+#source ~/.sirfrc")
+endif()
+message("To set up all the environmental variables for SIRF, just type: source ~/.sirfrc")
+
+configure_file(env_ccppetmr.csh.in ${CCPPETMR_INSTALL}/bin/env_ccppetmr.csh)
```
Comment (casperdcl) [inline SuperBuild.cmake]: surely we also need:

```cmake
configure_file(env_ccppetmr.sh.in ${CCPPETMR_INSTALL}/bin/env_ccppetmr.sh)
```

**Comment by KrisThielemans** [comment]
@casperdcl, we don't want to append to .bashrc, as discussed elsewhere. it wouldn't work for anyone who deson't run bash anyway.

**Comment by casperdcl** [comment]
I was referring to appending `. env_ccppetmr.sh` to `.bashrc`, and `source env_ccppetmr.csh` for `.cshrc` etc

**Comment by KrisThielemans** [comment]
ok. we can instruct the user to do that. let's not do it as part of our superbuild.

**Comment by paskino** [comment]
I agree that the test should be **also** on SIRF. 

Now ${PYTHON_EXECUTABLE} is used as suggested by @KrisThielemans

**Comment by paskino** [comment]
I’ve managed to use TEST_AFTER_INSTALL for the external project SIRF in order to use the tests of SIRF rather than in to the SuperBuild as @KrisThielemans [suggested](https://github.com/CCPPETMR/SIRF-SuperBuild/pull/32#issuecomment-329603373). 

However, it does not do what we want for a number of reasons: 

1. For the tests to run properly Gadgetron must run
2. If Gadgetron is built, it’s built as last project, so it won’t be running during the tests
3. We may build Gadgetron before SIRF but we still need to find the way to make Gadgetron run.
4. If Gadgetron runs, the tests PASS
5. If the tests fail (because Gadgetron doesn’t run—very likely) the build fails

I guess that what we want is that the SIRF tests pass. We’ve agreed –I believe— that the user should be responsible to run Gadgetron. So the only viable way I see is that we keep the tests both in the Superbuild and in SIRF. (Actually the Superbuild tests just [link](https://github.com/CCPPETMR/SIRF-SuperBuild/blob/ctest/SuperBuild.cmake#L101) to the test files in SIRF.)

So the ‘make test’ phase IMO must happen in the SuperBuild in 3 steps :

1. Build the SuperBuild
2. Start Gadgetron
3. make test

Do you see any other options?

**Comment by bathomas** [comment]
From what I've read, it seems that if you trigger a process (e.g. ```gadgetron```) from CMake via ```execute_process```, then it will wait (or timeout) for a response. This isn't useful to us as we want to keep ```gadgetron``` alive until we've run the test. However, maybe we could build a small application/script that forks and detaches ```gadgetron``` and returns a status value, then run the tests and kills ```gadgetron```.  In principle on Windows you could (in C++) mimic the behaviour of ```fork()``` with ```CreateProcess()```. Any thoughts? Too much work?

**Comment by KrisThielemans** [comment]
Seems too much effort. @paskino suggested order

1.    Build the SuperBuild
2.   Start Gadgetron
3.   make test

seems fine (with additional step of source the env files). In some sense, it's a more complete test: does the Gadgetron config work. It also prevents trouble on Windows where we cannot build Gadgetron anyway.

**Comment by paskino** [comment]
If we are happy with the 3 steps solution, we could merge the PR.

**Comment by KrisThielemans** [comment]
somewhat confused now. Once `ctest` is enabled for SIRF, do we need this here? If we use `TEST_COMMAND` (as opposed to `TEST_AFTER_INSTALL`, wouldn't we be able to follow the suggested workflow, while not having to duplicate the test here?

**Comment by paskino** [comment]
IMO the tests should be both in the SuperBuild and in SIRF. Imagine you are a user, you build the superbuild and then `make test`. Imagine you are a developer, you change something in SIRF and then you `make test`.

**Comment by KrisThielemans** [comment]
yes, agreed. but we cannot have duplicate code. So, we have them in SIRF, and somehow tell the superbuild do to "make test" in SIRF when you do "make test" in the SuperBuild. (same for gadgetron etc of course). I thought `TEST_COMMAND` might do the trick.

Alternative option might be to add a target to the SuperBuild CMake called `test_SIRF` and let that do its stuff.

**Comment by KrisThielemans** [comment]
perfect.

can you then add something in the README.md and also modify .travis.yml to use this as opposed to the current manual lines.

**Comment by paskino** [comment]
The build fails with this log message 

```
-- Installing: /home/travis/build/CCPPETMR/SIRF-SuperBuild/INSTALL/share/gadgetron/chroot/nvidia-copy.sh
[ 97%] [34m[1mCompleted 'Gadgetron'[0m
[ 97%] Built target Gadgetron
make: *** [all] Error 2

travis_time:end:283247ae:start=1506335531288183999,finish=1506336568000020844,duration=1036711836845
[0K
[31;1mThe command "make -j 6" failed and exited with 2 during .[0m

Your build has been stopped.

```

Sincerely I don't see what failed.

**Comment by bathomas** [comment]
Probably related to line 84 of .travis.yml:
```
 - make -j 6
```
My guess is this consumes too many resources are stops. I'd suggest amending the line to just ```make``` and see if this fixes the problem.

**Comment by KrisThielemans** [comment]
might then time-out, but worth a try.

you can also just rerun the job to see if it goes through 2nd time.

**Comment by paskino** [comment]
normally I use -j2 it speeds up download time and it's not too resource hungry.
Currently building single thread.

File: .travis.yml [hunk 2]
```diff
@@ -81,15 +81,14 @@ install:
...
   - pip install --only-binary=numpy,scipy,matplotlib numpy scipy matplotlib
   - ./cmake-*/bin/cmake $BUILD_FLAGS $EXTRA_BUILD_FLAGS .
-  - make -j 6
+  - make
   - mv INSTALL/share/gadgetron/config/gadgetron.xml.example INSTALL/share/gadgetron/config/gadgetron.xml
   - source $PWD/INSTALL/bin/env_ccppetmr.sh
   - #env
 
 script:
   - python ./SIRF/src/xSTIR/pSTIR/tests/test1.py
```
Comment (KrisThielemans) [inline .travis.yml]: should this not be deleted?

File: .travis.yml [hunk 3]
```diff
@@ -81,15 +81,14 @@ install:
...
   - ./INSTALL/bin/gadgetron >& gadgetron.log&
-  - python ./SIRF/src/xGadgetron/pGadgetron/tests/fully_sampled.py
-  - python ./SIRF/src/xGadgetron/pGadgetron/tests/undersampled.py
+  - ctest
```
Comment (KrisThielemans) [inline .travis.yml]: I like your `-verbose` option. let's add it here.

File: README.md [hunk 4]
```diff
@@ -53,6 +54,75 @@ source INSTALL/bin/env_ccppetmr.csh
...
+
+The user may also run the SIRF tests independently of the SuperBuild. Just enter the SIRF build directory and launch ctest:
+
+```bash
+cd SIRF-prefix/src/SIRF-build
+ctest
+```
+## Build of specific versions
+   
+The SuperBuild allows the user to change the versions of the projects it's building. This is done at the configuration stage that happens when you issue the `cmake` command in the root of the Superbuild.
```
Comment (KrisThielemans) [inline README.md]: split line.

Note that `cmake` doesn't have to be run in the root. It's cleaner to do out-of-source builds
```
git clone ...
mkdir build
cd build
cmake ../SIRF-SuperBuild
```
so, adjust text here (I'd just remove the "in the root of the SuperBuild"), and probably add a note on the instructions elsewhere

File: README.md [hunk 5]
```diff
@@ -53,6 +54,75 @@ source INSTALL/bin/env_ccppetmr.csh
...
+
+```bash
+cd SIRF-prefix/src/SIRF-build
+ctest
+```
+## Build of specific versions
+   
+The SuperBuild allows the user to change the versions of the projects it's building. This is done at the configuration stage that happens when you issue the `cmake` command in the root of the Superbuild. 
+
+There is a `DEVEL_BUILD` tag that allows to build the upstream/master versions of all packages
```
Comment (KrisThielemans) [inline README.md]: insert something like "By default, the SuperBuild will build the latest stable release of SIRF and associated versions of the dependencies."

Also, can we somehow say that the table contains hashes etc for release 0.9.0?

File: README.md [hunk 6]
```diff
@@ -53,6 +54,75 @@ source INSTALL/bin/env_ccppetmr.csh
...
+|`ISMRMRD_TAG` | `35012c6c8000616546c2d6b1757eba0c5b21b2d4` | `master` |
+
+To use the `DEVEL_BUILD` option one may (on the terminal)
+
+```bash
+
+cmake . -DDEVEL_BUILD=ON
+```
+
+Additionally one may want to use only a specific version of a package. This is achieved by adding the right tag to the command line (see the table above for available tags):
```
Comment (KrisThielemans) [inline README.md]: split line and add something like. "Note that the CMake options in the table are Advanced Options. When running the CMake GUI (or `ccmake`) they will therefore only be visible when you toggle those on."

File: SuperBuild/External_SIRF.cmake [hunk 7]
```diff
@@ -82,6 +82,7 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
 	INSTALL_DIR ${SIRF_Install_Dir}
     DEPENDS
         ${${proj}_DEPENDENCIES}
+    TEST_AFTER_INSTALL 0
```
Comment (KrisThielemans) [inline SuperBuild/External_SIRF.cmake]: delete?

File: env_ccppetmr.csh.in [hunk 8]
```diff
@@ -11,3 +11,6 @@ setenv LD_LIBRARY_PATH @CCPPETMR_INSTALL@/lib:$LD_LIBRARY_PATH
 setenv DYLD_FALLBACK_LIBRARY_PATH @CCPPETMR_INSTALL@/lib:$DYLD_FALLBACK_LIBRARY_PATH
 setenv PYTHONPATH @CCPPETMR_INSTALL@/python:$PYTHONPATH
 set path=( $path @CCPPETMR_INSTALL@/bin )
+
+# set nano as editor
```
Comment (KrisThielemans) [inline env_ccppetmr.csh.in]: we cannot do this here. we should only do SIRF stuff, not our preferences. (This is for the VM only)

File: env_ccppetmr.sh.in [hunk 9]
```diff
@@ -18,3 +18,6 @@ PYTHONPATH=@CCPPETMR_INSTALL@/python:$PYTHONPATH
 export PYTHONPATH
 
 PATH=$PATH:@CCPPETMR_INSTALL@/bin
+
```
Comment (KrisThielemans) [inline env_ccppetmr.sh.in]: as above

File: README.md (hunk 4 above)
Comment (paskino) [reply README.md]: but issuing `make` must be done in the SIRF-SuperBuild directory, or not?

Comment (KrisThielemans) [reply README.md]: no. `make` is always where `cmake` created the build files (i.e. the Makefiles in this case)

**Comment by KrisThielemans** [comment]
hi. presumably failing because of sourceforge outage now. But in any case, I think there's a few comments for the README outstanding?

**Comment by evgueni-ovtchinnikov** [comment]
SIRF branch sirf_sptr just built on VM without conflicts with new STIR. (It builds with the old one too.)

**Comment by KrisThielemans** [comment]
sourceforge still down. I did merge SIRF ctest so this will actually get tested once it's up again. But there are a few comments for the README outstanding :-;

**Comment by KrisThielemans** [comment]
3 jobs ok. The 3rd them failed in the `ctest` step. 
```
  - python: 3.5
    env: EXTRA_BUILD_FLAGS="-DUSE_SYSTEM_HDF5=ON -DUSE_SYSTEM_FFTW3=ON" CC=gcc-5 CXX=g++-5
  - python: 2.7
    env: EXTRA_BUILD_FLAGS="-DUSE_SYSTEM_SWIG=ON -DUSE_SYSTEM_HDF5=ON -DUSE_SYSTEM_FFTW3=ON" CC=gcc-5 CXX=g++-5
  - python: 3.5
    env: EXTRA_BUILD_FLAGS="-DUSE_SYSTEM_HDF5=OFF -DUSE_SYSTEM_FFTW3=OFF -DDEVEL_BUILD=ON" CC=gcc-5 CXX=g++-5
  - python: 2.7
env: EXTRA_BUILD_FLAGS="-DUSE_SYSTEM_SWIG=OFF -DUSE_SYSTEM_HDF5=OFF -DUSE_SYSTEM_FFTW3=OFF -DDEVEL_BUILD=ON" CC=gcc-5 CXX=g++-5
```
Here's the log. https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/280751611/config

Build succeeded but all 3 tests failed. I don't think it did anything at all with Gadgetron, so it's probably a python config thing. No idea why.

This brings us to getting useful output from the ctest when run on Travis. Possibly there's something useful in Testing/Temporary/LastTest.log. That might sit in this case inside the SIRF-build. Anyone any ideas on how to get that out from Travis (or typed out to the screen if the test failed)?

**Comment by paskino** [comment]
Yes, let's add a cat LastTest.log

**Comment by paskino** [comment]
The reason of it failing is 
`ImportError: No module named 'numpy'`

**Comment by KrisThielemans** [comment]
@casperdcl , I guess this has to do Python environments and pip not being the same env as python. can you help?

**Comment by casperdcl** [comment]
It also says
```
Command: "/usr/bin/python3.4" "test1.py"
```

Which is wrong as it should be:

```
Command: "/home/travis/virtualenv/python3.5.3/bin/python" "test1.py"
```

as this is the python environment installed and activated by travis

**Comment by paskino** [comment]
Should the environment variables be sourced?
It seems that in the install section there is `source $PWD/INSTALL/bin/env_ccppetmr.sh`, but there's no one in the script section.

**Comment by KrisThielemans** [comment]
as far as I understand, if you source once, it holds then for the rest of the travis script (the "install" and "script" etc sections are (I think) only used to give you a nice log). In any case, this is not relevant to this  problem: we don't set anything python specific in the `env_ccppetmr.*sh` scripts (and gadgetron etc would have failed to run).

**Comment by KrisThielemans** [comment]
I think @casperdcl  is correct that CMake's
```
include(FindPythonInterp)
```
tries to use its own logic and not necessarily honours the virtualenv. There's some reports in the internet about  this I believe (but I haven't researched this). it's apparently especially a problem on MacOSX (which we are currently already working around). see e.g. https://github.com/pybind/pybind11/issues/99. sigh.

It'd need some investigation of the cmake module to see if we can do something like
'''
cmake -DPYTHON_EXECUTABLE:FILEPATH=${PYTHON_EXECUTABLE}`
```
If so, we'd have to pass this frm the Superbuild to SIRF.

This is not an issue with this branch though. I therefore suggest that @paskino fixes the remaining README issues and that we merge.

**Comment by KrisThielemans** [comment]
a 2014 thread with a suggestion
https://public.kitware.com/pipermail/cmake/2014-August/058368.html

rather amazing problem, but might be worth checking if it isn't fixed a more recent CMake.

**Pull request merged**